http/vertx/src/main/java/io/cloudevents/http/vertx/CeVertx.java

```diff
@@ -0,0 +1,137 @@
+                .putHeader(HttpHeaders.createOptimized(EVENT_TYPE_KEY), HttpHeaders.createOptimized(ce.getEventType()))
+                .putHeader(HttpHeaders.createOptimized(SOURCE_KEY), HttpHeaders.createOptimized(ce.getSource().toString()))
+                .putHeader(HttpHeaders.createOptimized(EVENT_ID_KEY), HttpHeaders.createOptimized(ce.getEventID()));
+
+        // read optional headers
+        ce.getEventTypeVersion().ifPresent(eventTypeVersion -> {
+            request.putHeader(HttpHeaders.createOptimized(EVENT_TYPE_VERSION_KEY), HttpHeaders.createOptimized(eventTypeVersion));
+        });
+
+        ce.getEventTime().ifPresent(eventTime -> {
+            request.putHeader(HttpHeaders.createOptimized(EVENT_TIME_KEY), HttpHeaders.createOptimized(eventTime.format(DateTimeFormatter.ISO_OFFSET_DATE_TIME)));
+        });
+
+        ce.getSchemaURL().ifPresent(schemaUrl -> {
+            request.putHeader(HttpHeaders.createOptimized(SCHEMA_URL_KEY), HttpHeaders.createOptimized(schemaUrl.toString()));
+        });
+
+        ce.getData().ifPresent(data -> {
+            request.write(data.toString());
+        });
+    }
+
+    public static CloudEvent<?> readFromRequest(final HttpServerRequest request) {
+
+        final MultiMap headers = request.headers();
+        final CloudEventBuilder builder = new CloudEventBuilder();
+
+        // just check, no need to set the version
+        readRequiredHeaderValue(headers, CLOUD_EVENTS_VERSION_KEY);
+
+        builder
+                // set required values
+                .eventType(readRequiredHeaderValue(headers, EVENT_TYPE_KEY))
+                .source(URI.create(readRequiredHeaderValue(headers ,SOURCE_KEY)))
+                .eventID(readRequiredHeaderValue(headers, EVENT_ID_KEY))
+
+                // set optional values
+                .eventTypeVersion(headers.get(EVENT_TYPE_VERSION_KEY))
+                .contentType(headers.get(HttpHeaders.CONTENT_TYPE));
+
+                final String eventTime = headers.get(EVENT_TIME_KEY);
+                if (eventTime != null) {
+                    builder.eventTime(ZonedDateTime.parse(eventTime, DateTimeFormatter.ISO_OFFSET_DATE_TIME));
+                }
+
+                final String schemaURL = headers.get(SCHEMA_URL_KEY);
+                if (schemaURL != null) {
+                    builder.schemaURL(URI.create(schemaURL));
+                }
+
+
+        // get the extensions
+        final Map<String, String> extensions =
+                headers.entries().stream()
+                        .filter(header -> header.getKey().startsWith(HEADER_PREFIX))
+                        .collect(Collectors.toMap(h -> h.getKey(), h -> h.getValue()));
+
+        builder.extensions(extensions);
+        request.bodyHandler(buff -> {
+            if (buff.length()>0) {
+                builder.data(buff.toJsonObject().toString());
```

Comment:
@vietj @jponge this does read the body. However the actual `bodyhandler` is finished, after this method returns.

The HTTP headers got read, but the actual data is not inside the returned Cloudevent. 

see: https://github.com/cloudevents/sdk-java/pull/14/files#diff-36cb369b019aba77d5754e83bf85c583R67

Reply:
it is not clear what you mean, in the body handler you should get the full buffered body and that's actually the end of the HttpServerRequest lifecycle.

Reply:
> yes readFromRequest will be returned before the body handler is called because it is event driven. if you have code after readFromRequest you should pass it into an handler to the readFromRequest method so the body handler can read it.

@vietj ideally the `readFromRequest` returns _after_ the body handler, so that it does contain both: headers and the body, not just only the headers ;-)

Not exactly sure, atm how to do that

Reply:
@vietj We discussed offline with @matzew, I'm sure it will be resolved soon ;-)